File: tank.py
from actions import Backward, Forward, Left, Right


class Tank:

  DEFAULT_DIRECTION = 'N'
  DEFAULT_POSITION = (0, 0)

  def __init__(self):
    self.direction = Tank.DEFAULT_DIRECTION
    self.position = Tank.DEFAULT_POSITION
    self.actions = {
      'b': Backward(),
      'f': Forward(),
      'l': Left(),
      'r': Right(),
    }

  def execute(self, orders):
    for order in orders:
      action = self.actions[order]
      self.direction = action.turn(self.direction)
      self.position = action.move(self.position)

  def where_am_i_going(self):
    return self.direction

  def where_i_am(self):
    return self.position

File: tank.ts
import { Action, Position } from './actions';
import { Backward, Forward, Left, Right } from './actions';
import { North } from './directions';

export class Tank {

  private static DEFAULT_DIRECTION = new North();
  private static DEFAULT_POSITION = [0, 0] as Position;

  private direction: string;
  private position: Position;

  private actions: {[order: string]: Action};

  constructor() {
    this.direction = Tank.DEFAULT_DIRECTION;
    this.position = Tank.DEFAULT_POSITION.slice() as Position;

    this.actions = {
      'b': new Backward(),
      'f': new Forward(),
      'l': new Left(),
      'r': new Right(),
    };
  }

  public execute(orders: string[]): void {
    for (const order of orders) {
      const action = this.actions[order];
      this.direction = action.turn(this.direction);
      this.position = action.move(this.position);
    }
  }

  public where_am_i_going(): string {
    return this.direction.toString();
  }

  public where_i_am(): Position {
    return this.position;
  }
}
